rules/okta_rules/okta_potentially_stolen_session.py

```diff
@@ -0,0 +1,83 @@
+from ast import literal_eval
+from datetime import datetime, timedelta
+from difflib import SequenceMatcher
+
+from panther_base_helpers import deep_get, okta_alert_context
+from panther_oss_helpers import get_string_set, put_string_set, set_key_expiration
+
+PREVIOUS_SESSION = {}
+FUZZ_RATIO_MIN = 0.95
+
+
+def rule(event):
+    # pylint: disable=global-statement
+    # ensure previous session info is avaialable in the alert_context for investigation
+    global PREVIOUS_SESSION
+
+    # Filter only on app access and session start events
+    if (
+        event.get("eventType") == "user.authentication.sso"  # user opened app
+        or event.get("eventType") == "user.session.start"  # user loged in
+    ):
+
+        # lookup if we've previously stored the session cookie
+        PREVIOUS_SESSION = get_string_set(
+            deep_get(event, "authenticationContext", "externalSessionId"),
+        )
+
+        # For unit test mocks we need to eval the string to a set
+        if isinstance(PREVIOUS_SESSION, str):
+            PREVIOUS_SESSION = literal_eval(PREVIOUS_SESSION)
+
+        # if the session cookie was seen before
+        if PREVIOUS_SESSION is not None:
+            # we use a fuzz match to compare the current and prev user agent.
+            # We cannot do a direct match since Okta can occasionally maintain
+            # a session across browser upgrades.
+
+            # the user-agent is the longest string in the set
+            prev_ua = [x for x in PREVIOUS_SESSION if len(x) > 25]
+            diff_ratio = SequenceMatcher(
+                None, deep_get(event, "client", "userAgent", "rawUserAgent", default=""), prev_ua[0]
+            ).ratio()
+
+            # is this session being used from a new ASN and a different browser
+            if (
+                str(deep_get(event, "securityContext", "asNumber", default=""))
+                not in PREVIOUS_SESSION
+                and diff_ratio < FUZZ_RATIO_MIN
+            ):
+                # make the fuzz ratio available in the alert context
+                PREVIOUS_SESSION.add("Fuzz Ratio: " + str(diff_ratio))
+                return True
+
+        # If the sessionID has not been seen before, store information about it
+        else:
+            key = deep_get(event, "authenticationContext", "externalSessionId")
```

Comment:
update: I suspect counter against something like `sessionid  + event['client']['userAgent']` may be preferable

Reply:
I think we may need to change direction slightly. Unfortunately, we are going to run into a constraint in mocking the tests if we make two calls to `get_string_set` in the code. We can only mock one possible return value per function name in the test so I cannot return the correct value.

Due to this, I am going to move to a tagging method, where we prepend "user_agent" to the raw user agent when storing it in the set. This will guarantee we can find that string when we search for the user agent during comparison.